sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_custom_operation.asciidoc

```diff
@@ -0,0 +1,184 @@
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Backend support status
+
+This extension is currently implemented in {dpcpp} only for GPU devices and
+only when using the CUDA or HIP backends.  Attempting to use this extension in
+kernels that run on other devices or backends may result in undefined
+behavior.  Be aware that the compiler is not able to issue a diagnostic to
+warn you if this happens.
+
+
+== Overview
+
+This extension is derived from an experimental AdaptiveCpp extension, which is
+documented
+https://github.com/sycl_ext_oneapi/sycl_ext_oneapi/blob/develop/doc/enqueue-custom-operation.md[here]._
+
+The extension aims to integrate asynchronous interop work submitted through the
+backend API with the asynchrony of the SYCL runtime DAG. As such it is required
+that any interop calls used within `sycl_ext_oneapi_enqueue_custom_operation`
+are asynchronous.
+
+```c++
+sycl::queue q;
+
+q.submit([&](sycl::handler &cgh) {
+    sycl::accessor acc{buf, cgh};
+
+    cgh.sycl_ext_oneapi_enqueue_custom_operation([=](sycl::interop_handle &h) {
+      // Can extract device pointers from accessors
+      void *native_mem = h.get_native_mem<sycl::backend::hip>(acc);
+      // Can extract stream (note: get_native_queue() may not be 
+      // supported on CPU backends)
+      hipStream_t stream = h.get_native_queue<sycl::backend::hip>();
+      // Can extract HIP device (note: get_native_device() may not be
+      // supported on CPU backends)
+      int dev = h.get_native_device<sycl::backend::hip>();
+      // Can enqueue arbitrary backend operations. This could also be a kernel
+      // launch or call to a library that enqueues operations on the stream etc
+      hipMemcpyAsync(target_ptr, native_mem, test_size * sizeof(int),
+                      hipMemcpyDeviceToHost, stream);
+    });
+  });
+q.wait();
+```
+
+Provided that the correct dependencies are handled in the SYCL runtime, there
+is no need to do any further synchronization within the backend API in the user
+submitted lambda.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ENQUEUE_CUSTOM_OPERATION` to one of the values defined
+in the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+_And follow the text with a table like this *unless the extension is
+"experimental"*.  Note that your table may have more than one row if it
+has multiple versions._
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Synchronization
```

Comment:
The synchronization seems underspecified.  Here are the things I would want to know as a user of this extension:

* When is the lambda to `ext_oneapi_enqueue_custom_operation` executed?  Is it executed synchronously by the thread that calls `ext_oneapi_enqueue_custom_operation`, or is it called when some dependency is satisfied?  If so, what dependency?

* Precisely when does the event that is returned from `queue::submit` become "complete"?  I understand that you want it to be complete when the asynchronous work inside the lambda finishes, but these seems imprecise.  For example, what if the lambda submits async work to a completely different CUDA stream?  I wonder if the lambda should return a native CUDA event, which indicates when the custom operation completes.

Reply:
@AerialMantis had some ideas on this as well

Reply:
I have updated the section to add more information. A further review would be great, thanks

Reply:
So for (1), I agree, I don't think we should be too prescriptive here, we need to leave freedom for different implementation approaches. I think the important thing we need to specify is that the function is called exactly once, and is called when commands are enqueued to the backend, after any dependent commands have been enqueued, but without having to wait for dependent commands to complete execution.

Depending on the implementation, it could be executed immediately in the calling thread or deferred and called by a separate runtime thread. This is also important when we consider SYCL graphs in the future as in that case this function would be called during graph finalisation.

This will be a little tricky to define when it comes to a KHR as the SYCL specification currently doesn't make a distinction between a command submitted to the SYCL runtime and then the command submitted to the backend, so we may need to introduce new terminology for this, but I think for a vendor extension we can simply describe the semantics for this specific interface.

For (2) if a custom command is enqueued to a submit which returns an event I think the event should represent the completion of the native commands submitted. Since the interface itself doesn't include the creation or propagation of an event, instead a single backend stream or queue is provided to enqueue commands to, if a SYCL event is required then the implementation should record an event to the stream or queue used after the function completes and return this.

If the user was to use another stream or queue than the one provided via the interop API, this should be undefined behaviour as there would be no way for the SYCL runtime to track this. Without as you described having an API which returns a native event. We had this feature in our previous host task extension, however, we concluded this was not strictly necessary, made the interface more verbose and in some cases prevented the SYCL runtime from optimising away the use of native events.

Reply:
The latest draft is much better than before, and I think I understand the intent now.  I think the wording could still be improved, though.  In particular, it currently reads like a outline of possible implementations.  It should instead focus on the required semantics.  I'd also prefer to write this using our latest style, where there is an _Effects_ paragraph(s) that explain what the API does.  Here's a first draft:

> ```
> template <typename Func>
> void ext_oneapi_enqueue_custom_operation(Func&& interopCallable);
> ```
> _Constraints:_ The `Func` must a C++ callable object which takes a single parameter of type `interop_handle`.
>
> _Effects:_ The `interopCallable` object is called exactly once, and this call may be made asynchronously even after the calling thread returns from `ext_oneapi_enqueue_custom_operation`.
>
> The call to `interopCallable` may submit one or more asynchronous tasks to the native backend object that underlies the queue associated with this handler, and these tasks become a SYCL command that is added to the queue.  If the enclosing command group has any dependencies, these become implicit dependencies of that command.  This happens, for example, if the command group calls `handler::depends_on` or if it construct an accessor.  Note, however, that the asynchronous tasks submitted by `interopCallable` are the SYCL command, not the call to `interopCallable`.  The call to `interopCallable` may happen even before the dependencies are satisfied.
>
> The SYCL command described above completes once all of the native asynchronous tasks it contains have completed.
>
> The call to `interopCallable` must not submit any synchronous tasks to the native backend object, and it must not block waiting for any tasks to complete.  The call also must not add tasks to backend objects that underly any other queue, aside from the queue that is associated with this handler.  If it does any of these things, the behavior is undefined.

I'm still not convinced that we need to give the implementation the freedom to call `interopCallable` asynchronously.  If we require it to be called synchronously, applications can capture variables by reference, and everything will still work.  I think we could also remove the restriction on adding synchronous tasks from the function.  Doing so would have well-defined (though perhaps weird) semantics.  Is there a reason to allow this flexibility aside from graphs?

I'm not sure I buy the argument about needing to delay the call to `interopCallable` when recording a graph.  If you call the function each time the graph is finalized, you will end up calling it more than once, and this also violates the semantics.  Would it be possible to call `interopCallable` once (synchronously) and then record the native tasks it adds?

Reply:
>I'm still not convinced that we need to give the implementation the freedom to call interopCallable asynchronously. If we require it to be called synchronously, applications can capture variables by reference, and everything will still work.

So I can see the benefit of requiring `interopCallable` be called synchronously for capturing variables by reference, but this would limit flexibility in implementations, and would effectively enforce an implicit flush to the backend whenever the API is used. This isn't necessarily an issue for DPC++ as enqueue APIs call directly to the backend, so perhaps we could have it be synchronous for this vendor extension, but when we move to a KHR, we likely don't not want to have this limitation.

> I think we could also remove the restriction on adding synchronous tasks from the function. Doing so would have well-defined (though perhaps weird) semantics. Is there a reason to allow this flexibility aside from graphs?

When we specified this restriction, in general we were trying to avoid the user defining functions which would cause significant blocks or even deadlocks. If the `interopCallable` were to be called synchronously in the calling thread, this perhaps wouldn't be too bad as the user could infer exactly when the function would execute so they would be less likely to run into surprises, but if it's called asynchronously it becomes more difficult for the user to infer about when they will take the performance hit.

> I'm not sure I buy the argument about needing to delay the call to interopCallable when recording a graph. If you call the function each time the graph is finalized, you will end up calling it more than once, and this also violates the semantics. Would it be possible to call interopCallable once (synchronously) and then record the native tasks it adds?

For SYCL graphs, as I understand we typically only finalize a graph once, if there is a case where a graph is finalized more than once, we may have to consider how this is worded, as we would only want the function to be called once during graph creation. So we did consider the approach of having the `interopCallable` called when the backend is in a recording state, but there are a couple of issues with that. Firstly that this requires the backend to support recording/capturing of graphs and not all of them do, secondly this would enforce a particular implementation strategy for SYCL graphs where the graph is created as you go rather than when `finalize()` is called, which is not how it is currently implemented in DPC++.

Reply:
> For SYCL graphs, 

Hugh has put in an exception right now for using this extension with SYCL-Graphs. I'm not sure if specifying the semantics of the interaction of these two extensions if this is [intended to be merged by the end of the week](https://github.com/intel/llvm/pull/14136#issuecomment-2206382790). I'd suggest doing this as an isolated piece of follow on work. If we want to formalize that direction we can create an Issue section of the bottom of this spec to track it and also add an entry to https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc#712-interaction-with-other-extensions

>  if there is a case where a graph is finalized more than once

This is definitely possible, finalizing the same graph multiple times the only way to have a graph execute concurrently right now.

Reply:
I've updated this and the SYCL graphs doc to say that the ext shouldn't be used with SYCL graphs.

Reply:
The phrasing of the AdaptiveCpp extension is:

> This function object will be evaluated at DAG submission. 

See [here](https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/doc/enqueue-custom-operation.md). Perhaps this is a better phrasing than what we have here, which spells out that a SYCL RT may submit work eagerly or lazily. 

While it would be nice to capture things by reference in this lambda, the main purpose of `enqueue_custom_operation` is to do interop with the backend, so most objects that should be used within the user callable should be obtained through the interop handle. Objects that aren't obtained through the interop handle, like USM pointers, should be treated with the same caution as byref capture args, since they will likely be used by asynchronous backend calls, where the execution of such calls persists past the exit of the user submitted callable. As such I think there isn't a huge upside to having synchrony in the lambda call, given that the work sumbitted within the lambda is asynchronous.

Reply:
I don't think the phrase "This function object will be evaluated at DAG submission" is more clear.  In fact, I would interpret the phrase as meaning the lambda is executed synchronously, while you seem to think it means something else.

Regarding synchronous vs asynchronous and implementation freedom ... note that the core SYCL spec *does* say that the lambda passed to `queue::submit` is executed synchronously.  Therefore, I think there is a precedent for being somewhat proscriptive here.  Do we have some reason to believe that an implementation really would require the lambda passed to `ext_oneapi_enqueue_custom_operation` to be asynchronous, or is it more of an abstract argument about preserving implementation freedom?

So far, I think the strongest argument for asynchronous execution is the interaction with the graph API, but there seem to be a bunch of unresolved questions with this anyway.

Reply:
I think that the command group lambda being executed synchronously is OK to specify since it doesn't prescribe any particular execution model of the SYCL RT. i.e. even though the command group lambda is executed synchronously that doesn't mean that the command needs to be launched synchronously.

If we restrict this to say that the `enqueue_native_command` lambda must be executed synchronously, we are essentially saying that this extension is only useful for SYCL runtimes that use eager dispatch, like DPC++. This would necessitate another extension to get the same nice dovetailing of interop asynchrony with the SYCL runtime's asynchrony, if the SYCL runtime is to use lazy dispatch. Of course this would be silly and incomprehensible for users.

The key thing that this extension does is allow the user to dispatch interop work in the same manner as normal SYCL commands are dispatched by the SYCL runtime. Since the SYCL RT can choose whether to use eager or lazy dispatch then the spec should also allow this callable to be executed synchronously or asynchronously.

Reply:
I agree the command group function is slightly different, in that it must allow the capture of local variables, and those captured variables are of types with complicated lifetime implications, so it needs to be synchronous. However, for `enqueue_native_command` while I can see the value of capturing variables by reference, this is not the typical usage and we would actively discourage capturing variables of those types by reference.

So future-proofing for SYCL Graphs is a strong argument for not enforcing synchronous execution, but there are also implementations for regular SYCL code. The SYCL execution model doesn't prescribe how and when a backend will enqueue a command, to allow implementations freedom to optimise the scheduling and batching of commands. For example commands could be re-ordered based on dependencies or commands could be collected together and submitted all at once. If we were to introduce a requirement that the function provided to `enqueue_native_command` is executed synchronously that would prevent these kinds of optimisations.

As I understand DPC++ doesn't currently perform these kinds of optimisations, so this is more theoretical, though in the past I worked on a backend to ComputeCpp which took advantage of this freedom to gain better performance. I'm not sure the benefit here would be worth the limitation it would introduce, I would prefer to leave this as unspecified for now. With the expected usage of the feature, the functions themselves should generally be self contained and without any expectation of when they are executed, so I wouldn't expect this to have a significant impact on users.

Reply:
OK, I'm convinced that we should not require the lambda to be called synchronously.

I would still like my previous [comments](https://github.com/intel/llvm/pull/14136#discussion_r1664733723) above to be addressed.  The proposed text I suggest there is already written to assume that the lambda might be called asynchronously.  The things I would like to address by making this change are:

* Write the specification in a way that describes the required semantics.  Currently, I think it reads more like a list of possible ways the API could be implemented.
* Clarify the rules explaining when the command completes.  (The command is the thing that executes in the DAG according to its dependencies.)  I think my proposed wording is better, but I think it could be further clarified.  Proposals for further clarification are definitely welcome.
* Use the new style based on _Constraints_, _Effects_, etc. paragraphs, which is based on the C++ specification.

Reply:
Thanks Greg, sorry the above comment somehow got lost for me in this discussion. Thanks for linking. Will amend

Reply:
I've updated the docs to reflect your suggestions

Reply:
Thanks for adding this, but I should have been more clear.  I think this wording should be the formal description of the `ext_oneapi_enqueue_native_command` function.  Specifically, I'd like to see the document structured like this:

* Remove your existing sections "Synchronization" and "Additions to handler class"
* Add a new section "Additions to handler class".  Start that section with a sentence:

  > This extension adds the following new member function to the SYCL `handler` class:

  and follow that sentence with the synopsis and then the _Constraints_ and _Effects_ paragraphs that you currently have in "Overview".

* You should remove the sentence:

  >For more description of synchronization guarantees, see the Synchronization section.

  since that section won't exist anymore.  If there are more semantics of `ext_oneapi_enqueue_native_command` that are not already captured in the _Effects_ paragraphs, add more information to those paragraphs.  However, do not just copy the text you have in Synchronization.  I think that current text reads more like an outline of implementation possibilities.  To be honest, I think the current wording in _Effects_ already captures the requirements you list in "Synchronization".  But, if you think it is missing something, it should be added.

If you have general information you want to say about this extension, for example how `ext_oneapi_enqueue_native_command` differs from `host_task`, add this information to the Overview section.  The Overview is non-normative (i.e. not the official specification), and it is meant to introduce your extension and explain why it's useful.  The "Specification" section is the normative (i.e. official specification) part of the document.  Any implementation of the extension **must** behave according to what you say in the "Specification" section.  If you have additional examples that you think will help clarify the usage of the extension to your reader, consider adding an "Examples" section towards the bottom.  If you have an "Examples" section, it is a peer of "Specification", so the examples are not part of the official specification.

Reply:
Thanks Greg. I've made those changes now. The specification section is very brief, as most of the API is about usage rather than introducing new types etc. Let me know if you think this is OK.

Reply:
Move the "Additions to handler class" section under "Specification".  This is the right place for that section because it is the formal specification of `ext_oneapi_enqueue_native_command` and the "Specification" section is the formal part of the document.  The outline should look like this:

* Notice
* Contact
* Dependencies
* Status
* Backend support status
* Overview
* Specification
  * Feature test macro
  * Additions to handler class
* Issues

When deciding what to put under "Specification" think of it this way.  Could an implementor implement this extension purely by reading the Specification section?

Reply:
Thanks. Have reordered sections. I think the `Additions to handler class` describes everything necessary for an implementer to implement this. Let me know if you think this is OK